### DISCUSSION TITLE
External binary dependencies? (os/exec) fork/exec

### DISCUSSION THREAD

**Discussion by prologic** [containers/podman#15308] [Q&A] [answered]
Hi Podman devs 👋

Does podman depend on any external binaries and thus the use of the [os/exec](https://pkg.go/dev/os/exec) package, essentially fork/exec for some functionality?

Docker for example sadly depends on `iptables` on the system, which makes running Docker in restricted environments or embedded systems (e.g: Busybox, Custom Linux) quite difficult.

Kind regards

**Comment by mheon** [comment, marked as answer]
We have a number of such dependencies. The OCI runtime is an obvious one - Podman does not directly launch containers, but instead makes use of `crun`, `runc`, or another OCI runtime. Further, we do not monitor the OCI runtime ourselves, as Podman is not a daemon; instead, we use [conmon](github.com/containers/conmon), which is an extremely lightweight daemon that wraps a single container and monitors its status. Rootless Podman needs external, OS-provided setuid binaries (`newuidmap` and `newgidmap`), though I believe it can function in a (seriously reduced) state without them. Networking is also configured externally (via either the CNI plugins, or Netavark in newer Podman), and as this needs to modify the system firewall and iptables does not have an API, that will be fork/execing `iptables` to accomplish that (though I'm working on an nftables driver for Netavark that will remove that particular need, at least).

I'm probably forgetting some here, but suffice to say we make use of a lot of external utilities.

**Reply by prologic** [reply]
Thanks for your detailed reply! 👌 Helps answers a lot of my questions!
